libs/types/src/investments.rs

```diff
@@ -52,3 +57,210 @@ where
+		}
+	}
+}
+
+/// The collected investment/redemption amount for an account
+#[derive(Encode, Default, Decode, Clone, Eq, PartialEq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct CollectedAmount<Balance: Default + MaxEncodedLen> {
+	/// The amount which was collected
+	/// * If investment: Tranche tokens
+	/// * If redemption: Payment currency
+	pub amount_collected: Balance,
+
+	/// The amount which was converted during processing based on the
+	/// fulfillment price(s)
+	/// * If investment: Payment currency
+	/// * If redemption: Tranche tokens
+	pub amount_payment: Balance,
+}
+
+/// A representation of an investment identifier and the corresponding owner.
+///
+/// NOTE: Trimmed version of `InvestmentInfo` required for foreign investments.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, Default, TypeInfo, MaxEncodedLen)]
+
+pub struct ForeignInvestmentInfo<AccountId, InvestmentId, TokenSwapReason> {
+	pub owner: AccountId,
+	pub id: InvestmentId,
+	pub last_swap_reason: Option<TokenSwapReason>,
+}
+
+/// A simple representation of a currency swap.
+#[derive(
+	Clone,
+	Default,
+	Copy,
+	PartialOrd,
+	Ord,
+	PartialEq,
+	Eq,
+	Debug,
+	Encode,
+	Decode,
+	TypeInfo,
+	MaxEncodedLen,
+)]
+pub struct Swap<
+	Balance: Clone + Copy + EnsureAdd + EnsureSub + Ord + Debug,
+	Currency: Clone + PartialEq,
+> {
+	/// The incoming currency, i.e. the desired one.
+	pub currency_in: Currency,
+	/// The outgoing currency, i.e. the one which should be replaced.
+	pub currency_out: Currency,
+	/// The amount of outgoing currency which shall be exchanged.
+	pub amount: Balance,
+}
+
+impl<Balance: Clone + Copy + EnsureAdd + EnsureSub + Ord + Debug, Currency: Clone + PartialEq>
+	Swap<Balance, Currency>
+{
+	/// Ensures that the ingoing and outgoing currencies of two swaps...
+	/// * Either match fully (in1 = in2, out1 = out2) if the swap direction is
+	///   the same for both swaps, i.e. (into pool, into pool) or (into foreign,
+	///   into foreign)
+	/// * Or the ingoing and outgoing currencies match (in1 = out2, out1 = in2)
+	///   if the swap direction is opposite, i.e. (into pool, into foreign) or
+	///   (into foreign, into pool)
+	pub fn ensure_currencies_match(
+		&self,
+		other: &Self,
+		is_same_swap_direction: bool,
+	) -> DispatchResult {
+		if is_same_swap_direction
+			&& (self.currency_in != other.currency_in || self.currency_out != other.currency_out)
+		{
+			Err(DispatchError::Other(
+				"Swap currency mismatch for same swap direction",
+			))
+		} else if !is_same_swap_direction
+			&& (self.currency_in != other.currency_out || self.currency_out != other.currency_in)
+		{
+			Err(DispatchError::Other(
+				"Swap currency mismatch for opposite swap direction",
+			))
+		} else {
+			Ok(())
+		}
+	}
+}
+
+/// A representation of an executed investment decrement.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, Default, TypeInfo, MaxEncodedLen)]
+
+pub struct ExecutedForeignDecreaseInvest<Balance, Currency> {
```

Comment:
Nit - why not have this and the following 2 as enums?

Reply:
If I understand you correctly to refactor this into
```rust
enum ExecutedForeign {
	DecreaseInvest { .. },
	CollectInvest { .. },
	CollectRedeem { .. }
}
```

.. then I don't see usability advantages as this would require the hooks which expect these explicit types as `Status` type to short circuit on mismatching enum variants, i.e. `DecreasedForeignInvestOrderHook` for variants `ExecutedForeign::{CollectInvest, CollectRedeem}`. I think this adds unnecessary obfuscation